### ISSUE TITLE
comet {autoexit, false} problem in Firefox

### ISSUE THREAD

**Issue by viplifes** [ostinelli/misultin#10]
I create a comet application

checked the option {autoexit, false} in all browsers

it does not work in firefox. connection is not closed

This problem is a browser? (firefox 3.6.15, firefox 4.0)

**Comment by ostinelli** [comment]
yes, firefox does not close sockets unless you really close all of the browser.

vipfiles, this is not a support panel, it's an issue report. if you find issues you're welcome to drop them here, however please use this section wisely. thank you.

r.

**Comment by wjlroe** [comment]
When this happens - you close the tab of the websocket in Firefox, you get a stacktrace - 
    http process <0.71.0> has died with reason: {function_clause,
                                             [{lists,reverse,[none]},
                                              {misultin_websocket,
                                               handle_data,6},
                                              {misultin_http,handle_data,8}]}, removing from references of open connections

Couldn't this be handled more gracefully to close the connection server-side? Or at least send the closed message to the websocket pid?

**Comment by ostinelli** [comment]
this issue was about comet, not about websockets. however, this should be indeed the intended behaviour on websockets too.

i've just tried the examples:

misultin_websocket_example.erl
misultin_websocket_event_example.erl
misultin_websocket_example_ssl.erl

and all three worked as expected, with no "has died with reason". could you please tell me which code you're using so that i can reproduce and fix what you are reporting?

**Comment by wjlroe** [comment]
My code is modelled on misultin_websocket_event_example.erl.
I copied that example into my app and started it from the supervisor and tried it.
It crashed like this:

```
=ERROR REPORT==== 24-Mar-2011::09:42:37 ===
    module: misultin
    line: 290
http process <0.63.0> has died with reason: {function_clause,
                                         [{lists,reverse,[none]},
                                          {misultin_websocket,
                                           handle_data,6},
                                          {misultin_http,handle_data,8}]}, removing from references of open connections

=ERROR REPORT==== 24-Mar-2011::09:42:37 ===
Error in process <0.63.0> on node 'poker1@127.0.0.1' with exit value: {function_clause,[{lists,reverse,[none]},{misultin_websocket,handle_data,6},{misultin_http,handle_data,8}]}
```

When I openned it from Firefox then closed the tab. 

So looking at handle_data/6 it looks like this could be hacked to send close to the Pid when (on line 243), the socket is none - i.e. it's died. What do you think? I'm surprised you can't reproduce it though. Perhaps there's another issue?

Many thanks,
Will

**Comment by ostinelli** [comment]
definitely can't reproduce it. what exactly do you mean with "I copied that example into my app and started it from the supervisor and tried it"?

which app? which supervisor? please try to start it as standalone and see if you still experience this error, otherwise can you please strip down your code to the minimum which can reproduce this error.

**Comment by wjlroe** [comment]
OK sorry that was really unclear. So basically I mashed the gen_server websocket example and the event example. 

So I've tried using the misultin_websocket_event_example module in examples/ and running it thus:

```
will@Byrd ~/mycode/poker master % erl -pa deps/misultin/ebin/ ebin/                                                                 
Erlang R13B04 (erts-5.7.5) [source] [smp:2:2] [rq:2] [async-threads:0] [hipe] [kernel-poll:false]

Eshell V5.7.5  (abort with ^G)
1> misultin_websocket_event_example:start(9876).
{ok,<0.34.0>}
Receive from Pid: <0.40.0>
2> 
2> 
=ERROR REPORT==== 24-Mar-2011::10:30:48 ===
        module: misultin
        line: 290
http process <0.39.0> has died with reason: {function_clause,
                                             [{lists,reverse,[none]},
                                              {misultin_websocket,
                                               handle_data,6},
                                              {misultin_http,handle_data,8}]}, removing from references of open connections

=ERROR REPORT==== 24-Mar-2011::10:30:48 ===
Error in process <0.39.0> with exit value: {function_clause,[{lists,reverse,[none]},{misultin_websocket,handle_data,6},{misultin_http,handle_data,8}]}
```

**Comment by ostinelli** [comment]
please state FireFox version, as the stable version 3.x does not support websockets, and haven't checked FF 4 specs.

also, the io:format "Receive from Pid: <0.40.0>" message is not in the example.

**Comment by wjlroe** [comment]
I added that io:format to check on the Pid basically. 

Firefox 4.0. You have to unblock the websocket in about:config
https://developer.mozilla.org/en/WebSockets

Have you tried reproducing this with Firefox?

**Comment by ostinelli** [comment]
see https://github.com/ostinelli/misultin/pull/11

**Comment by wjlroe** [comment]
Nice. Thanks!
